protocols/gossipsub/src/handler.rs

```diff
@@ -325,7 +325,8 @@ impl ConnectionHandler for Handler {
 
             // If there was a fatal error, close the connection.
             if let Some(error) = reported_error {
-                return Poll::Ready(ConnectionHandlerEvent::Close(error));
+                self.keep_alive = KeepAlive::No;
+                log::info!("Gossipsub error: {error}");
```

Comment:
I think this indirection is confusing.

We should "handle" the error as soon as possible. Now that we are changing the error handling from returning it to logging, we can do that as soon as the error occurs instead of storing it in a variable (`upgrade_errors`) and delaying the handling.

For the specific state where the protocol is unsupported, we can store only that state and return the specific event within `poll`.

Those that make sense?

Reply:
Clear.
Still, it bugs me that there is no return statement. Because further processing seems like extra unnecessary work. This statement applies to all places where `return` is replaced by a report.

Reply:
> Still, it bugs me that there is no return statement. Because further processing seems like extra unnecessary work.

Yes, we shouldn't even be doing any of this if we can just log the error in the callback where we are first encountering it.

The original reason why this variable exists is because we can't do anything in the callback where it happens.

`ConnectionHandler` is a state machine and `poll` "advances" it. Thus, if errors happen in other parts we need to sort-of buffer them and return them on the next "tick" (i.e. `poll`).

Reply:
A decision must be taken on how to return. Please see my other comment [below](https://github.com/libp2p/rust-libp2p/pull/3625#discussion_r1140067178)